File: algorithms/java/Huffman.java
import java.util.*;
import java.io.*;

public class Huffman {

    static class Node implements Comparable<Node> {
        int  freq;
        byte ch;
        Node left, right;
        Node(int f, byte c)           { freq=f; ch=c; }
        Node(int f, Node l, Node r)   { freq=f; left=l; right=r; }
        public int compareTo(Node o)  { return Integer.compare(freq, o.freq); }
    }

    static int[] codeLens = new int[256];

    static void buildCodes(Node n, int depth) {
        if (n == null) return;
        if (n.left == null && n.right == null) {
            codeLens[n.ch & 0xFF] = depth > 0 ? depth : 1;
            return;
        }
        buildCodes(n.left,  depth+1);
        buildCodes(n.right, depth+1);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br   = new BufferedReader(new InputStreamReader(System.in));
        String         text = br.readLine().trim();
        int            n    = text.length();

        int[] freq = new int[256];
        for (char c : text.toCharArray()) freq[c]++;

        long start = System.nanoTime();

        PriorityQueue<Node> pq = new PriorityQueue<>();
        int unique = 0;
        for (int i = 0; i < 256; i++) {
            if (freq[i] > 0) {
                pq.offer(new Node(freq[i], (byte)i));
                unique++;
            }
        }

        while (pq.size() > 1) {
            Node l = pq.poll(), r = pq.poll();
            pq.offer(new Node(l.freq + r.freq, l, r));
        }

        if (!pq.isEmpty()) buildCodes(pq.poll(), 0);

        long totalBits = 0;
        for (int i = 0; i < 256; i++)
            if (freq[i] > 0) totalBits += (long)freq[i] * codeLens[i];

        double ms    = (System.nanoTime() - start) / 1e6;
        double ratio = (double)totalBits / (8.0 * n);

        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        pw.println(unique);
        pw.println(totalBits);
        pw.printf("%.2f%n", ratio);
        pw.flush();
        System.err.printf("%.3f%n", ms);
    }
}

File: algorithms/julia/huffman.jl
using DataStructures

mutable struct HNode
    freq  :: Int
    ch    :: UInt8
    left  :: Union{HNode, Nothing}
    right :: Union{HNode, Nothing}
end

HNode(f, c) = HNode(f, c, nothing, nothing)

Base.isless(a::HNode, b::HNode) = a.freq < b.freq

function build_codes!(node::HNode, depth::Int, codes::Vector{Int})
    if isnothing(node.left) && isnothing(node.right)
        codes[node.ch + 1] = depth > 0 ? depth : 1
        return
    end
    isnothing(node.left)  || build_codes!(node.left,  depth+1, codes)
    isnothing(node.right) || build_codes!(node.right, depth+1, codes)
end

function huffman(text::String)
    freq = zeros(Int, 256)
    for c in text; freq[UInt8(c)+1] += 1; end

    pq = PriorityQueue{HNode, Int}()
    unique = 0
    for (i, f) in enumerate(freq)
        if f > 0
            n = HNode(f, UInt8(i-1))
            enqueue!(pq, n, f)
            unique += 1
        end
    end

    while length(pq) > 1
        l = dequeue!(pq)
        r = dequeue!(pq)
        p = HNode(l.freq + r.freq, 0x00, l, r)
        enqueue!(pq, p, p.freq)
    end

    codes = zeros(Int, 256)
    isempty(pq) || build_codes!(dequeue!(pq), 0, codes)

    total = sum(freq[i] * codes[i] for i in 1:256 if freq[i] > 0)
    return unique, total
end

function main()
    text = strip(read(stdin, String))
    n    = length(text)

    # JIT warmup
    huffman("abcd")

    start        = time_ns()
    unique, total = huffman(text)
    ms           = (time_ns() - start) / 1e6

    ratio = total / (8.0 * n)
    println(unique)
    println(total)
    println(@sprintf("%.2f", ratio))
    println(stderr, ms)
end

using Printf
main()